taichi/runtime/cuda/kernel_launcher.cpp

```diff
@@ -62,8 +63,9 @@ void KernelLauncher::launch_llvm_kernel(Handle handle,
           device_buffers[i] = arg_buffers[i];
         }
         // device_buffers[i] saves a raw ptr on CUDA device.
-        ctx.set_struct_arg({i, 1}, (uint64)device_buffers[i]);
-
+        ctx.set_ndarray_ptrs(
+            i, (uint64)device_buffers[i],
+            (uint64)ctx.array_ptrs[{i, TypeFactory::GRAD_PTR_POS_IN_NDARRAY}]);
```

Comment:
Do we plan to support grad with numpy and Pytorch arrays? If so, IIUC the grad ptrs need to be processed like the primal ptrs as above.

Reply:
ditto~